nifi-nar-bundles/nifi-questdb-bundle/nifi-questdb/src/main/java/org/apache/nifi/questdb/embedded/EmbeddedClient.java

```diff
@@ -0,0 +1,152 @@
+import io.questdb.cairo.sql.RecordCursor;
+import io.questdb.cairo.sql.RecordCursorFactory;
+import io.questdb.griffin.CompiledQuery;
+import io.questdb.griffin.SqlCompiler;
+import io.questdb.griffin.SqlCompilerFactoryImpl;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.mp.SCSequence;
+import io.questdb.mp.TimeoutBlockingWaitStrategy;
+import org.apache.nifi.questdb.Client;
+import org.apache.nifi.questdb.DatabaseException;
+import org.apache.nifi.questdb.InsertRowDataSource;
+import org.apache.nifi.questdb.QueryResultProcessor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+final class EmbeddedClient implements Client {
+    private final static Logger LOGGER = LoggerFactory.getLogger(EmbeddedClient.class);
+
+    private final Supplier<CairoEngine> engine;
+    private final AtomicBoolean disconnected = new AtomicBoolean(false);
+
+    EmbeddedClient(final Supplier<CairoEngine> engine) {
+        this.engine = engine;
+    }
+
+    @Override
+    public void execute(final String query) throws DatabaseException {
+        checkConnectionState();
+
+        try (final SqlCompiler compiler = getCompiler()) {
+            final CompiledQuery compile = compiler.compile(query, getSqlExecutionContext());
+            compile.execute(new SCSequence(new TimeoutBlockingWaitStrategy(5, TimeUnit.SECONDS)));
+        } catch (final SqlException | CairoError e) {
+            throw new DatabaseException(e);
+        }
+    }
+
+    @Override
+    public void insert(
+        final String tableName,
+        final InsertRowDataSource rowDataSource
+    ) throws DatabaseException {
+        checkConnectionState();
+
+        if (!rowDataSource.hasNextToInsert()) {
+            LOGGER.debug("No rows to insert into {}", tableName);
+            return;
+        }
+
+        final TableToken tableToken = engine.get().getTableTokenIfExists(tableName);
+
+        if (tableToken == null) {
+            throw new DatabaseException(String.format("Table Token for table [%s] not found", tableName));
+        }
+
+        try (
+            final TableWriter tableWriter = engine.get().getWriter(tableToken, "adding rows")
+        ) {
+            final TableWriterBasedInsertRowContext context = new TableWriterBasedInsertRowContext(tableWriter);
+
+            while (rowDataSource.hasNextToInsert()) {
+                context.addRow(rowDataSource);
+            }
+
+            LOGGER.debug("Committing {} rows", tableWriter.getRowCount());
+            tableWriter.commit();
+        } catch (final Exception | CairoError e) {
+            // CairoError might be thrown in extreme cases, for example when no space left on the disk
+            LOGGER.error("Add rows to table [{}] failed", tableName, e);
+            throw new DatabaseException(e);
```

Comment:
Is there a reason for logging the error and throwing the exception? The caller should also catch and log the exception.